### PR TITLE
Doxygen header templates

### PR DESCRIPTION
**Pull request by MatthewMasarik-NOAA**
# Pull Request Summary
Acceptance of doxygen headers to be used for documentation of WW3 Fortran source code units. 

## Description
Provide a detailed description of what this PR does.
* Adds a markdown file (`./docs/ww3_doxy_tmpl.md`) which will serve as a reference for required header tags when marking up WW3 source for doxygen.
* A `docs/` folder is also added which will eventually hold the documentation.
* An initial version of the doxygen config file, `Doxygen.in`, is also included in `docs/`.

What bug does it fix, or what feature does it add?
* N/A.

Is a change of answers expected from this PR?
* No.  No changes to source code at this time.

Mention any labels that should be added:
  *  _documentation_.

### Issue(s) addressed
Please list any issues associated with this PR, including those the PR will fix/close. 
- fixes noaa-emc/ww3/issues/[#547](https://github.com/NOAA-EMC/WW3/issues/547).

### Commit Message
ww3_doxy_tmpl.md: add doxygen header templates

### Check list  
- [x] Branch is up to date with the authoritative repository (NOAA-EMC) develop branch. 
- [x] Checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop). 
- [na] If a version number update is required, checked the [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) checklist. 
- [na] If a new feature was added, a regression test for testing the new feature is added. 
 
### Testing

* How were these changes tested?
  * No testing -- there were no source code changes.
* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)
  * No, see answer above.
* Have the matrix regression tests been run (if yes, please note HPC and compiler)?
  * N/A.
* Please provide the summary output of matrix.comp (_matrix.Diff.txt_, _matrixCompFull.txt_ and _matrixCompSummary.txt_):    
  * N/A.
* Please indicate the expected changes in the regression test output ([Note the known list of non-identical tests](https://github.com/NOAA-EMC/WW3/wiki/How-to-use-matrix.comp-to-compare-regtests-with-master#4-look-at-results)).
  * N/A.
* Please provide the summary output of matrix.comp (_matrix.Diff.txt_, _matrixCompFull.txt_ and _matrixCompSummary.txt_):
  * N/A.

### CODE REVIEW

File: docs/ww3_doxy_tmpl.md
Comment (captainkirk99) [inline docs/ww3_doxy_tmpl.md:49]: Did you forget the author and date tags for the function?

File: docs/ww3_doxy_tmpl.md
Comment (captainkirk99) [inline docs/ww3_doxy_tmpl.md:37]: The brief tag is not needed for subroutines/functions. THe first sentence will be automatically used as the brief description.

Brief tags are required at the module and file level - autobrief does not work for files and modules, just subroutines and functions.

File: docs/ww3_doxy_tmpl.md
Comment (captainkirk99) [inline docs/ww3_doxy_tmpl.md:55]: We do not doxygenate in-line comments (i.e. comments in the code itself), except when adding documentation for variables.

So we might have:

```
integer:: i !< Some doxygen description of i.

! Some code comment that is not doxygen.

```

**Review by JessicaMeixner-NOAA — APPROVED** [review]

**Comment by JessicaMeixner-NOAA** [comment]
@aliabdolali what are we waiting for to merge this PR in?

**Comment by aliabdolali** [comment]
> @aliabdolali what are we waiting for to merge this PR in?

I needed time to do my final review.

**Review by aliabdolali — APPROVED** [review]

File: docs/ww3_doxy_tmpl.md
Comment (edwardhartnett) [inline docs/ww3_doxy_tmpl.md:8]: Don't put the file name in the file.

Over time, as files get moved around, these will become incorrect. However, if never specified, doxygen will always use the correct file name.

So the top of our files look like this:

```
!>    @file
!>    @brief Pack up a Grid Definition Section (Section 3) and add it to
!>    a GRIB2 message.
!>    @author Stephen Gilbert @date 2000-05-01


```

File: docs/ww3_doxy_tmpl.md
Comment (edwardhartnett) [inline docs/ww3_doxy_tmpl.md:18]: The details tag is never needed and should not be used.

Doxygen is set up to allow a lot of different ways to specify the docs. It's possible, for example, to have the docs completely separate from the code files. And in this case, some of these extra tags may be needed.

But in our case, we are using doxygen in the simplest and most programmer friendly way. Whatever is not needed, should not be asked for, giving programmers the easiest and simplest path to good documentation habits.

So the details tag should be removed everywhere in this document. It's not needed.

File: docs/ww3_doxy_tmpl.md
Comment (edwardhartnett) [inline docs/ww3_doxy_tmpl.md:37]: On subroutines and functions, the brief tag is not needed. The first sentence is automatically used as the brief description.

Again, this is to make the documentation less cluttered, and make it easier for programmers to properly document their functions without extra work.

So for functions and subroutines, we don't use either the brief or the details tags. Note in the following example, the first sentence will be the brief description, and the second would be considered detailed description.

```
!>    This subroutine is to extract arbitrary size values from a
!>    packed bit string, right justifying each value in the unpacked
!>    array without skip and iterations.
!>
!>    This should be used when input array IN has only one element. If IN
!>    has more elements, use G2_SBYTESC().
!>
!>    @param[in] IN Array input.
!>    @param[out] IOUT Unpacked array output.
!>    @param[in] ISKIP Initial number of bits to skip.
!>    @param[in] NBYTE Number of bits of each integer in IN to take.
!>
!>    @author Stephen Gilbert @date 2004-04-27
SUBROUTINE G2_GBYTEC(IN,IOUT,ISKIP,NBYTE)
  character*1 in(*)
  integer iout(*)
  CALL G2_GBYTESC(IN,IOUT,ISKIP,NBYTE,0,1)
  RETURN
END SUBROUTINE G2_GBYTEC

```

File: docs/ww3_doxy_tmpl.md
Comment (MatthewMasarik-NOAA) [reply docs/ww3_doxy_tmpl.md:8]: I was not aware doxygen pulled the name itself.  I'll make this change.  The PR just got accepted, so I'll discuss with Ali at a meeting we have later today to discuss how to add this in.  I'll resolve the conversation once it's done.

File: docs/ww3_doxy_tmpl.md
Comment (edwardhartnett) [inline docs/ww3_doxy_tmpl.md:58]: The !< marker can be used when var documentation is short enough to fit on one line, but is not mandatory. Just as with other doxygen documentation blocks, the !> marker can be used, and the documentation placed before the variable. 

It should be noted that all module and type variables need documentation. That is, each variable must be documented, none skipped (or you will get warnings).

Here's an example using the !< marker, where each element of documentation fits on one line:

```
module ip_grid_mod
  use ip_grid_descriptor_mod
  implicit none

  integer, public, parameter :: EQUID_CYLIND_GRID_ID_GRIB1 = 0 !< Integer grid number for equidistant cylindrical grid in grib1
  integer, public, parameter :: MERCATOR_GRID_ID_GRIB1 = 1 !< Integer grid number for Mercator grid in grib1
  integer, public, parameter :: LAMBERT_CONF_GRID_ID_GRIB1 = 3 !< Integer grid number for Lambert Conformal grid in grib1
  integer, public, parameter :: GAUSSIAN_GRID_ID_GRIB1 = 4 !< Integer grid number for Gaussian grid in grib1
  integer, public, parameter :: POLAR_STEREO_GRID_ID_GRIB1 = 5 !< Integer grid number for polar stereo grid in grib1
  integer, public, parameter :: ROT_EQUID_CYLIND_E_GRID_ID_GRIB1 = 203 !< Integer grid number for rotated equidistant cylindrical E-stagger grid
  integer, public, parameter :: ROT_EQUID_CYLIND_B_GRID_ID_GRIB1 = 205 !< Integer grid number for rotated equidistant cylindrical B-stagger grid

  integer, public, parameter :: EQUID_CYLIND_GRID_ID_GRIB2 = 0 !< Integer grid number for equidistant cylindrical grid in grib2
  integer, public, parameter :: ROT_EQUID_CYLIND_GRID_ID_GRIB2 = 1 !< Integer grid number for rotated equidistant cylindrical grid in grib2
  integer, public, parameter :: MERCATOR_GRID_ID_GRIB2 = 10 !< Integer grid number for Mercator grid in grib2
  integer, public, parameter :: POLAR_STEREO_GRID_ID_GRIB2 = 20 !< Integer grid number for polar stereo grid in grib2
  integer, public, parameter :: LAMBERT_CONF_GRID_ID_GRIB2 = 30 !< Integer grid number for Lambert conformal grid in grib2
  integer, public, parameter :: GAUSSIAN_GRID_ID_GRIB2 = 40 !< Integer grid number for Gaussian grid in grib2

```

Here's an example where more detail is required, and the documentation elements usually come before the variables:

```
  type gribfield
     integer :: version        !< GRIB edition number (currently 2).

     !>     Message Discipline (see [Code Table 0.0]
     !>    (https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_table0-0.shtml)).
     integer :: discipline

     !>    Contains the entries in the Identification Section (Section
     !>    1).
     !>    - idsect(1) Identification of originating Centre
     !>    (see Common Code Table C-1) 7 US National Weather Service. ([Table 0]
     !>    (https://www.nco.ncep.noaa.gov/pmb/docs/on388/table0.html)).
     !>    - idsect(2) Identification of originating Sub-centre ([Table C]
     !>    (https://www.nco.ncep.noaa.gov/pmb/docs/on388/tablec.html)).
     !>    - idsect(3) GRIB Master Tables Version Number. ([Table 1.0]
     !>    (https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_table1-0.shtml))
     !>    0 Experimental; 1 Initial operational version number
     !>    - idsect(4) GRIB Local Tables Version Number ([Table 1.1]
     !>    (https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_table1-1.shtml)).
     !>     - 0 Local tables not used
     !>     - 0 1-254 Number of local tables version used
     !>    - idsect(5) Significance of Reference Time ([Table 1.2]
     !>    (https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_table1-1.shtml)).
     !>     - 0 Analysis
     !>     - 1 Start of forecast
     !>     - 2 Verifying time of forecast
     !>     - 3 Observation time
     !>    - idsect(6) Year (4 digits)
     !>    - idsect(7) Month
     !>    - idsect(8) Day
     !>    - idsect(9) Hour
     !>    - idsect(10) Minute
     !>    - idsect(11) Second
     !>    - idsect(12) Production status of processed data (see [Table 1.3]
     !>    (https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_table1-3.shtml))
     !>     - 0 Operational products
     !>     - 1 Operational test products
     !>     - 2 Research products
     !>     - 3 Re-analysis products
     !>    - idsect(13) Type of processed data ([Table 1.4]
     !>    (https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_table1-4.shtml))
     !>     - 0 Analysis products
     !>     - 1 Forecast products
     !>     - 2 Analysis and forecast products
     !>     - 3 Control forecast products
     !>     - 4 Perturbed forecast products
     !>     - 5 Control and perturbed forecast products
     !>     - 6 Processed satellite observations
     !>     - 7 Processed radar observations
     integer, pointer, dimension(:) :: idsect => null()

     !>    Number of elements in idsect (always 13).
     integer :: idsectlen

     !>    Pointer to character array containing contents
     !>    of Local Section 2, if included.
     character(len = 1), pointer, dimension(:) :: local => null()

     !>    Length of array local.
     integer :: locallen


```

File: docs/ww3_doxy_tmpl.md
Comment (MatthewMasarik-NOAA) [reply docs/ww3_doxy_tmpl.md:18]: I will leave this call up to the code managers.  I had left these in because for me they gave some visual order when scanning through the code.  I do acknowledge they're functionally unnecessary though, so that case is understood.  I'm fine either way.

File: docs/ww3_doxy_tmpl.md
Comment (edwardhartnett) [reply docs/ww3_doxy_tmpl.md:58]: In general, training Fortran programmers on doxygen, I have found it important to tell them that not all comments will be converted to doxygen. Only the comment blocks outlined in this document. Regular comments in the code are not converted to doxygen. So not all comments in the code will be doxygen comments.

For example, note this subroutine starts with a doxygen comment block (with '!>') but the comments in the code remain simple Fortran comments (with just a '!').


```
!>    This subroutine finalizes a GRIB message after all grids and
!>    fields have been added. It adds the End Section ("7777") to the
!>    end of the GRIB message and calculates the length and stores it in
!>    the appropriate place in Section 0. This routine is used with
!>    routines gribcreate(), addlocal(), addgrid(), and addfield() to
!>    create a complete GRIB2 message.
!>
!>    @param[inout] cgrib Character array to contain the GRIB2 message.
!>    @param[in] lcgrib Maximum Length (in bytes) of array cgrib.
!>    @param[out] lengrib Length of the final GRIB2 message in octets
!>    (bytes).
!>    @param[out] ierr Error return code.
!>    - 0 no error.
!>    - 1 GRIB message was not initialized. call routine gribcreate() first.
!>    - 2 GRIB message already complete.
!>    - 3 Sum of Section byte counts doesn't add to total byte count.
!>    - 4 Previous Section was not 7.
!>
!>    @author Stephen Gilbert @date 2000-05-02
subroutine gribend(cgrib, lcgrib, lengrib, ierr)

  character(len = 1), intent(inout) :: cgrib(lcgrib)
  integer, intent(in) :: lcgrib
  integer, intent(out) :: lengrib, ierr

  character(len = 4), parameter :: grib = 'GRIB', c7777 = '7777'
  character(len = 4):: ctemp
  integer iofst, ibeg, lencurr, len

  ierr = 0

  !     Check to see if beginning of GRIB message exists
  ctemp = cgrib(1) // cgrib(2) // cgrib(3) // cgrib(4)
  if (ctemp .ne. grib) then
     print *, 'gribend: GRIB not found in given message.'
     ierr = 1
     return
  endif

  !     Get current length of GRIB message
  call g2_gbytec(cgrib, lencurr, 96, 32)

  !     Check to see if GRIB message is already complete
  !     ctemp=cgrib(lencurr-3)//cgrib(lencurr-2)//cgrib(lencurr-1) &
  !          //cgrib(lencurr)
  !     if (ctemp.eq.c7777) then
  !       print *, 'gribend: GRIB message already complete.'
  !       ierr=2
  !       return
  !     endif

  !     Loop through all current sections of the GRIB message to
  !     find the last section number.
  len = 16                    ! Length of Section 0

```

File: docs/ww3_doxy_tmpl.md
Comment (MatthewMasarik-NOAA) [reply docs/ww3_doxy_tmpl.md:37]: I left the brief in on subroutines/functions for the same reason I had left the details tag in, to me they add some structure when scanning through the code.  They are not required though so I could easily understand removing them.  I'll leave this call up to the code managers.

Comment (JessicaMeixner-NOAA) [reply docs/ww3_doxy_tmpl.md:37]: @MatthewMasarik-NOAA I'll support whatever you decide is best.

File: docs/ww3_doxy_tmpl.md
Comment (JessicaMeixner-NOAA) [reply docs/ww3_doxy_tmpl.md:8]: The PR is not merged so you can still make changes. If it gets merged, you can just make a new PR.

Comment (JessicaMeixner-NOAA) [reply docs/ww3_doxy_tmpl.md:8]: Oh I take that back, it has been merged, so just make a new PR to change the file with anything you'd like to be changed.

File: docs/ww3_doxy_tmpl.md
Comment (MatthewMasarik-NOAA) [reply docs/ww3_doxy_tmpl.md:58]: Good point on the '!>' and '!<' markers.  Also, I did mention in the module header that ALL module variables need to be documented.

File: docs/ww3_doxy_tmpl.md
Comment (edwardhartnett) [inline docs/ww3_doxy_tmpl.md:65]: The todo and bug flags should not be used by any NOAA programmer. Instead, use GitHub issues to describe bugs and desired future work.

File: docs/ww3_doxy_tmpl.md
Comment (edwardhartnett) [inline docs/ww3_doxy_tmpl.md:32]: Actually the date field can be specified in any format. It's not ever really parsed as a date, it's just treated as text. So you could have:

`@date Sometime Last Winter
`
and it would work.

Sometimes in NCEPLIBS all we know is the year. ;-)

File: docs/ww3_doxy_tmpl.md
Comment (edwardhartnett) [reply docs/ww3_doxy_tmpl.md:37]: This decision has already been made and is working well. Is there some reason you guys want to be different from the other doxygen projects at NOAA?

Remove the unnecessary elements. Programmers will not thank you for adding extra requirements, which are not real. And having your programmers, and all future programmers, deal with the brief and details tag is just extra work and extra difference from the way the rest of the EMC are doing documentation.

Comment (JessicaMeixner-NOAA) [reply docs/ww3_doxy_tmpl.md:37]: @edwardhartnett  I didn't know there was a NOAA wide mandate about use of doxygen. For WW3 we are really happy to start including doxygen coding updates and are super appreciative of the hard work @MatthewMasarik-NOAA has done to create these templates for WW3 to use.

File: docs/ww3_doxy_tmpl.md
Comment (MatthewMasarik-NOAA) [reply docs/ww3_doxy_tmpl.md:8]: Sounds good.  Will do.

File: docs/ww3_doxy_tmpl.md
Comment (MatthewMasarik-NOAA) [reply docs/ww3_doxy_tmpl.md:37]: I have a meeting already scheduled with Ali this afternoon that will give us a chance to discuss offline.

File: docs/ww3_doxy_tmpl.md
Comment (MatthewMasarik-NOAA) [reply docs/ww3_doxy_tmpl.md:32]: I agree, though the particular format chosen was what has been used in the existing WW3 code headers, so was meant to be consistent from that standpoint.

File: docs/ww3_doxy_tmpl.md
Comment (MatthewMasarik-NOAA) [reply docs/ww3_doxy_tmpl.md:65]: I get your point.  I will consider this in a new PR.

File: docs/ww3_doxy_tmpl.md
Comment (MatthewMasarik-NOAA) [reply docs/ww3_doxy_tmpl.md:37]: @edwardhartnett @JessicaMeixner-NOAA I'm putting together a new PR that will take these recommendations into consideration.  It will be posted this afternoon.

File: docs/ww3_doxy_tmpl.md
Comment (MatthewMasarik-NOAA) [reply docs/ww3_doxy_tmpl.md:8]: @edwardhartnett , please see new PR #593 .  This has been addressed.

File: docs/ww3_doxy_tmpl.md
Comment (MatthewMasarik-NOAA) [reply docs/ww3_doxy_tmpl.md:18]: I mulled the options and decided to leave these in.  I gave my thinking in the new PR #593 .

File: docs/ww3_doxy_tmpl.md
Comment (MatthewMasarik-NOAA) [reply docs/ww3_doxy_tmpl.md:37]: Posted:  PR #593.

File: docs/ww3_doxy_tmpl.md
Comment (MatthewMasarik-NOAA) [reply docs/ww3_doxy_tmpl.md:58]: This has been addressed.

File: docs/ww3_doxy_tmpl.md
Comment (MatthewMasarik-NOAA) [reply docs/ww3_doxy_tmpl.md:32]: This was left as is for consistency across the current WW3 code base.

File: docs/ww3_doxy_tmpl.md
Comment (MatthewMasarik-NOAA) [reply docs/ww3_doxy_tmpl.md:65]: Removed.

**Pull request merged**
